indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/SeekableStreamIndexTaskRunner.java

```diff
@@ -16,55 +16,1991 @@
+        // should resume to immediately finish kafka index task as failed
+        resume();
+        return Response.status(Response.Status.INTERNAL_SERVER_ERROR)
+                       .entity(Throwables.getStackTraceAsString(e))
+                       .build();
+      }
+      finally {
+        pauseLock.unlock();
+      }
+    }
+
+    resume();
+
+    return Response.ok(sequenceNumbers).build();
+  }
+
+  private void resetNextCheckpointTime()
+  {
+    nextCheckpointTime = DateTimes.nowUtc().plus(tuningConfig.getIntermediateHandoffPeriod()).getMillis();
+  }
+
+  @GET
+  @Path("/checkpoints")
+  @Produces(MediaType.APPLICATION_JSON)
+  public Map<Integer, Map<PartitionType, SequenceType>> getCheckpointsHTTP(@Context final HttpServletRequest req)
+  {
+    authorizationCheck(req, Action.READ);
+    return getCheckpoints();
+  }
+
+  private Map<Integer, Map<PartitionType, SequenceType>> getCheckpoints()
+  {
+    return new TreeMap<>(sequences.stream()
+                                  .collect(Collectors.toMap(
+                                      SequenceMetadata::getSequenceId,
+                                      SequenceMetadata::getStartOffsets
+                                  )));
+  }
+
+  /**
+   * Signals the ingestion loop to pause.
+   *
+   * @return one of the following Responses: 400 Bad Request if the task has started publishing; 202 Accepted if the
+   * method has timed out and returned before the task has paused; 200 OK with a map of the current partition sequences
+   * in the response body if the task successfully paused
+   */
+  @POST
+  @Path("/pause")
+  @Produces(MediaType.APPLICATION_JSON)
+  public Response pauseHTTP(
+      @Context final HttpServletRequest req
+  ) throws InterruptedException
+  {
+    authorizationCheck(req, Action.WRITE);
+    return pause();
+  }
 
   @VisibleForTesting
-  Response pause() throws InterruptedException;
+  public Response pause() throws InterruptedException
+  {
+    if (!(status == Status.PAUSED || status == Status.READING)) {
+      return Response.status(Response.Status.BAD_REQUEST)
+                     .entity(StringUtils.format("Can't pause, task is not in a pausable state (state: [%s])", status))
+                     .build();
+    }
+
+    pauseLock.lockInterruptibly();
+    try {
+      pauseRequested = true;
+
+      pollRetryLock.lockInterruptibly();
+      try {
+        isAwaitingRetry.signalAll();
+      }
+      finally {
+        pollRetryLock.unlock();
+      }
+
+      if (isPaused()) {
+        shouldResume.signalAll(); // kick the monitor so it re-awaits with the new pauseMillis
+      }
+
+      long nanos = TimeUnit.SECONDS.toNanos(2);
+      while (!isPaused()) {
+        if (nanos <= 0L) {
+          return Response.status(Response.Status.ACCEPTED)
+                         .entity("Request accepted but task has not yet paused")
+                         .build();
+        }
+        nanos = hasPaused.awaitNanos(nanos);
+      }
+ 
```

Comment:
Does this need to check `getRunnerType()`?

Reply:
is this needed because `exclusiveStartingPartitions` will not be empty for Kafka when initial offsets are fetched from metadata storage